### ISSUE TITLE
MAX: Ragdoll+ glitches

### ISSUE THREAD

**Issue by PaulFox12** [TheGameCreators/GameGuruRepo#2535]
See photos - Two character creator characters placed in a level are glitching in odd poses when killed. A bug with the ragdoll+ system? 
![death glitch 1](https://user-images.githubusercontent.com/78652980/158596232-533b132e-33ab-4c48-921a-08337cf30d64.png)
![death glitch 2](https://user-images.githubusercontent.com/78652980/158596234-92a0e090-4b67-44f9-8bbe-9b9250ad915b.png)

Below is the scene from the editor for layout context
![Scene](https://user-images.githubusercontent.com/78652980/158596551-63fd50ef-6fcf-4f04-95ee-4752c6f1c661.png)

Not seen here but also when killed characters will often cartwheel through the air several meters away before adopting their death pose - maybe looking for a clear area? Seems like Ragdoll+ needs looked at.

Really wish we could just use ragdoll as an option as many of the death animations repeat and often look quite naff. There's also no force considered if I kill a character at point blank range with a weapon like a shotgun for example. couple this with the above bugs and glitches and an option for normal ragdoll would be much preferred in many cases

**Comment by rvanner** [comment]
@PaulFox12 Can you send me the fpm of this level so we can test it?

**Comment by dllllllllllb** [comment]
@ZakJudges I think you have more experience with Characters, I haven't touched them yet :]

**Comment by ZakJudges** [comment]
Hi @PaulFox12, I wasn't able to reproduce the characters ending up in a pose like the ones seen in your screenshots. Do you still experience it in the latest build? :)

> Not seen here but also when killed characters will often cartwheel through the air several meters away before adopting their death pose
I didn't see any cartwheels, but I did see the characters sliding unnaturally to a different spot whilst performing their death anim.

@LeeBamberTGC See comments about characters moving before adopting their death pose. I've attached a map that I experienced the issue with :)
[ragdoll_plus_bug.zip](https://github.com/TheGameCreators/GameGuruRepo/files/8340239/ragdoll_plus_bug.zip)

**Comment by ZakJudges** [comment]
@LeeBamberTGC A surefire way to reproduce the wrong orientation of the pose is to kill a character whilst they are on stairs:
![ragdoll plus stairs](https://user-images.githubusercontent.com/32545391/159914511-dc3dbf6f-33f0-4b1b-a287-04a6d4703764.png)

**Comment by LeeBamberTGC** [comment]
@ZakJudges Yes the is happening because somehow the character has calculated that a death anim can be used, but the physics 'trick' to find the floor goes very wrong and spins the final orientation as the physics 'box trick' hits the banister and flips over. Not ideal as the trick was really meant for terrain and sloping floors.  I think a few more collision checks to go direct to ragdoll if they are in these types of enclosed spaces would be a good fix, for this and other similar tight spaces.

**Comment by LeeBamberTGC** [comment]
NOTE: I have assigned to you so you can check it out and learn how this system works. Eventually, I would like you to have a better grasp of the character AI, its physics system, the death anim system, and ragdoll.  This is a nice easy(ish) intro to completing a fix and learning the system at the same time :)

**Comment by ZakJudges** [comment]
@LeeBamberTGC Thanks, that all sounds exciting :)

I've added a fix for the soldiers ending up under the stairs. To fix it, I decreased fThresholdForMovingTooFarX and fThresholdForMovingTooFarZ. This seems to still allow the death anims to be found, but is a bit more strict about performing them when there isn't much room. Do you know if this might have any negative consequences elsewhere?

**Comment by LeeBamberTGC** [comment]
@ZakJudges Only way to know is to test all the demo levels that have ragdoll potential and keep an eye out. The solution sounds correct and what I might have done :)

**Comment by ZakJudges** [comment]
Hi @PaulFox12, could you confirm if the wrong orientation issue is now fixed for you in the most recent build?
We will be working on the ragdoll further in the future, so I will close this issue if so. Thanks :)
